shark/shark/src/main/java/shark/ApacheHarmonyInstanceRefReaders.kt

```diff
@@ -116,6 +118,37 @@ enum class ApacheHarmonyInstanceRefReaders : OptionalFactory {
     }
   },
 
+  // https://cs.android.com/android/platform/superproject/+/android-6.0.1_r81:libcore/luni/src/main/java/java/util/WeakHashMap.java
+  WEAK_HASH_MAP {
+    override fun create(graph: HeapGraph): VirtualInstanceReferenceReader? {
+      val weakHashMapClass = graph.findClassByName("java.util.WeakHashMap") ?: return null
+
+      // No table field in Apache Harmony impl.
+      val isOpenJdkImpl = weakHashMapClass.readRecordFields()
+        .any { weakHashMapClass.instanceFieldName(it) == "table" }
+
+      if (isOpenJdkImpl) {
+        return null
+      }
+
+      val hashMapClassId = weakHashMapClass.objectId
+
+      return InternalSharedHashMapReferenceReader(
```

Comment:
Unfortunately, we need more tweaking than this. WeakHashMap has weak keys. We need to reconnect this with how we define which references to follow or not.